# Assignment  Practice Text classification with Naive Bayes 

## Notebook made by   (If not filled in correctly: 0 pts for assignment)

__Name(s)__: 

__Student id(s)__ : 

### Pledge (taken from [Coursera's Honor Code](https://www.coursera.org/about/terms/honorcode) )



Put here a selfie with your photo where you hold a signed paper with the following text: (if this is team work, put two selfies here). The link must be to some place on the web, not to a local file. **Assignments without the selfies will not be graded and receive 0 points.**

> My answers to homework, quizzes and exams will be my own work (except for assignments that explicitly permit collaboration).

>I will not make solutions to homework, quizzes or exams available to anyone else. This includes both solutions written by me, as well as any official solutions provided by the course staff.

>I will not engage in any other activities that will dishonestly improve my results or dishonestly improve/hurt the results of others.

<img src='link to your selfie'/>

# Some example code

In [30]:
import pandas as pd

# Change to KVR1000.csv.gz if this becomes to0 slow for you
# normal KVR.csv.gz
# subset = KVR1000.csv
kvrdf= pd.read_csv('KVR.csv.gz', 
                   sep='\t', encoding='utf-8',
                   index_col=0, 
                   names=['jaar', 'partij','titel','vraag','antwoord','ministerie']) 
print (kvrdf.shape)
kvrdf.head()

(40516, 6)


jaar partij  \
KVR1000.xml     1994   PvdA   
 KVR10000.xml   1999   PvdA   
 KVR10001.xml   1999     SP   
 KVR10002.xml   1999   PvdA   
 KVR10003.xml   1999   PvdA   

                                                           titel  \
KVR1000.xml     De vragen betreffen de betrouwbaarheid van de...   
 KVR10000.xml   Vragen naar aanleiding van berichten (uitzend...   
 KVR10001.xml   Vragen naar aanleiding van de berichten "Nede...   
 KVR10002.xml   Vragen over de gebrekkige opvang van verpleeg...   
 KVR10003.xml   Vragen over onbetrouwbaarheid van filemeldingen.   

                                                           vraag  \
KVR1000.xml     Hebt u kennisgenomen van het televisieprogram...   
 KVR10000.xml   Kent u de berichten over de situatie in de Me...   
 KVR10001.xml   Kent u de berichten «Nederland steunt de Soeh...   
 KVR10002.xml   Kent u het bericht over onderzoek van Nu91 me...   
 KVR10003.xml   Hebt u kennisgenomen van de berichten over de...   

                                                        antwoord  \
KVR1000.xml     Ja. Het bedoelde geluidmeetpunt is eigendom v...   
 KVR10000.xml                                                      
 KVR10001.xml                                                      
 KVR10002.xml   Ja. Het onderzoek van NU’91 wijst uit dat het...   
 KVR10003.xml   Ja. Nee. Door de waarnemers van het Algemeen ...   

                                       ministerie  
KVR1000.xml                 Verkeer en Waterstaat  
 KVR10000.xml                            Justitie  
 KVR10001.xml                           Financien  
 KVR10002.xml   Volksgezondheid, Welzijn en Sport  
 KVR10003.xml               Verkeer en Waterstaat

In [98]:
kvrdf.ministerie.value_counts().head(10)

 Justitie (JUS)                                                   3219
 Volksgezondheid, Welzijn en Sport (VWS)                          2630
 Buitenlandse Zaken (BUZA)                                        1796
 Verkeer en Waterstaat (VW)                                       1441
 Justitie                                                         1333
 Sociale Zaken en Werkgelegenheid (SZW)                           1231
 Onderwijs, Cultuur en Wetenschappen (OCW)                        1187
 Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)     984
 Financiën (FIN)                                                   960
 Volksgezondheid, Welzijn en Sport                                 951
dtype: int64

In [99]:
import nltk
import string
from collections import Counter

# Join all titels of the questions together
allvragen = '\n'.join(str(v) for v in list(kvrdf.titel))

# remove punctuation
allvragen = allvragen.translate(str.maketrans({key: None for key in string.punctuation}))


print (len(allvragen), allvragen[:1000])

10075650  De vragen betreffen de betrouwbaarheid van de geluidsmetingen op Schiphol
 Vragen naar aanleiding van berichten uitzending NOVA 17 juli 1999 over onderzoeken van de Rijksrecherche waaruit blijt dat er twijfels bestaan over het professioneel gedrag van vier medewerkers van de kliniek
 Vragen naar aanleiding van de berichten Nederland steunt Soehartos fiscaal en Wij hebben een naam hoog te houden in NRC van 22 juli 1999
 Vragen over de gebrekkige opvang van verpleegkundigen met traumatische ervaringen door ziekenhuizen en zorginstellingen
 Vragen over onbetrouwbaarheid van filemeldingen
 Vragen naar aanleiding van het bericht in de Volkskrant van 7 juli 1999 dat het RIVM onjuiste gegevens zou hebben gehanteerd en gepresenteerd wat betreft de milieuaspecten van de Betuwelijn
 
 Vragen naar aanleiding van het artikel Eilander artsen vechten voor waarneming in de Leeuwarder Courant van 26 juni 1999 Deze vragen dienen ter aanvulling op eerdere vragen van de leden Van BlerckWoerdman

In [68]:
# Tokenize and turn into a bag of words
BoW = Counter(nltk.word_tokenize(allvragen))
# show the top 20 most used words 
BoW.most_common(20)

[('de', 105875),
 ('van', 98546),
 ('het', 48956),
 ('in', 36051),
 ('en', 30934),
 ('een', 27234),
 ('over', 22818),
 ('Vragen', 20820),
 ('op', 19408),
 ('dat', 18593),
 ('te', 17443),
 ('naar', 16878),
 ('voor', 15251),
 ('aanleiding', 12074),
 ('met', 11486),
 ('is', 11043),
 ('aan', 9982),
 ('die', 8845),
 ('zijn', 8565),
 ('vragen', 7987)]

In [70]:
from nltk.corpus import stopwords

DutchStop= stopwords.words('dutch')
print (len(DutchStop))

BoW= BoW= Counter([w for w in nltk.word_tokenize(allvragen) if not w in set(DutchStop)])
# show the top 20 most used words 
BoW.most_common(20)

101


[('Vragen', 20820),
 ('aanleiding', 12074),
 ('vragen', 7987),
 ('De', 5966),
 ('bericht', 4549),
 ('minister', 4328),
 ('artikel', 3269),
 ('Nederlandse', 3138),
 ('Nederland', 3033),
 ('nr', 2769),
 ('2007', 2746),
 ('eerdere', 2662),
 ('Indiener', 2562),
 ('Is', 2538),
 ('1', 2273),
 ('ingezonden', 2235),
 ('Betreft', 2111),
 ('ter', 2108),
 ('ten', 2097),
 ('onderzoek', 2094)]

# Practice Text classification with Naive Bayes  
        
        
        
<h3>Abstract</h3>
<p>We will do text classification on a collection of Dutch parliamentary questions.
    The website <a href="https://zoek.officielebekendmakingen.nl/zoeken/parlementaire_documenten">officielebekendmakingen.nl</a>lets you search in "kamervragen".
    You can donwload
    <a href='http://data.politicalmashup.nl/kamervragen/PoliDocs_Kamervragen.zip'>this zipfile with Kamervragen in XML</a>
    to see some of the  data in XML format. 
    It also contains style sheets to show the XML well in a browser.  
    The <a href='http://maartenmarx.nl/teaching/zoekmachines/LectureNotes/MySQL/'>MYSQL directory</a> contains an <a href='http://maartenmarx.nl/teaching/zoekmachines/LectureNotes/MySQL/KVR14807.xml'>example   Kamervraag XML file</a>. Note that in your browser you see the result of applying stylesheets. So choose View Source or open it in an editor.</p>

<h3>First exploration</h3>

See below.

<h2>Exercise</h2>

<p>We will use the fields in elements of the form <tt> &lt;item attribuut="Afkomstig_van"></tt> as our classes. 
    These are the ministeries to whom the question is addressed.
    An example is 
    <pre>
        &lt;item attribuut="Afkomstig_van">Landbouw, Natuurbeheer en Visserij (LNV)&lt;/item>
    </pre>
    Note that these labels are <strong>not normalized</strong>, see e.g. the counts below:
    <pre>
Justitie (JUS)                                                   3219
Volksgezondheid, Welzijn en Sport (VWS)                          2630
Buitenlandse Zaken (BUZA)                                        1796
Verkeer en Waterstaat (VW)                                       1441
Justitie                                                         1333
Sociale Zaken en Werkgelegenheid (SZW)                           1231
Onderwijs, Cultuur en Wetenschappen (OCW)                        1187
Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)     984
Financiën (FIN)                                                   960
Volksgezondheid, Welzijn en Sport                                 951
Economische Zaken (EZ)                                            946
Buitenlandse Zaken                                                753
Binnenlandse Zaken en Koninkrijksrelaties (BZK)                   725
Verkeer en Waterstaat                                             724
Defensie (DEF)                                                    646
Sociale Zaken en Werkgelegenheid                                  607
Landbouw, Natuurbeheer en Visserij (LNV)                          586
Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer            554
Onderwijs, Cultuur en Wetenschappen                               532
Vreemdelingenzaken en Integratie (VI)                             466
    </pre>
</p>

  <ol>
      <li>Normalize the values for "ministerie" and choose 10 ministeries to work with. </li>
      <li>Implement the two algorithms in Fig MRS.13.2, using your earlier code for creating term and document frequencies.
      It might be easier to use the representation and formula given in MRS section 13.4.1.</li>
      <li>On this collection, train NB text classifiers for 10 different classes with enough and interesting data.</li>
      <li>Compute for each term and each of your 10 classes its utility for that class using mutual information.</li>
      <li>For each class, show the top 10 words as in Figure 13.7 in MRS.</li>
      <li>Evaluate your classifiers using Precision, Recall and F1. (
           <br/>
          Give a table in which you show these values for using the top 10, top 100 terms and all terms, for all of your 10 classes.
          Thus do feature selection per class, and use for each class the top n best features for that class. 
          <br/>
      Also show the microaverage(s) for all 10 classes together.
      <br/>
      If you like you can also present this in a figure like MRS.13.8. 
      Then compute the F1 measure for the same number of terms as in that figure.</li>
      <li>Reflect and report briefly about your choices in this process and about the obtained results. </li>
  </ol>

<h3>Training/Testing</h3>
<p>It is important that you do not test your classifier using documents that have also been used in training.
    So split up your collection in a training set and a test set. A 80%-20% split is reasonable.

<br/>
    If you have too little data you can use 5 or <a href="http://en.wikipedia.org/wiki/Cross-validation_(statistics)#k-fold_cross-validation">10-fold cross validation</a>.</p>

<h2>Form of presentation</h2>
<ul>
    <li>Make slides or wikipages and have your system running and be able to accept documents from the web. </li>
    <li>Create one or two slides or wikipages for each of the sub exercises listed above.
</li>
<li>Make it clear in the heading of the slides which sub exercises you talk about.</li>
    <li>Show running code with one or two  good examples (a TP of course, but also a FP and an error-analysis is nice to show). </li>

</ul>

<h2>Form of handing in your final product</h2>
<ul>
    <li>An IPython notebook would be perfect, with clear indications which part of the code answers which subquestion.</li>
    <li>A clear git repo, with good comments and a clear separation and indication what code does what is also fine.</li>
    <li> You are free to program in whatever language you prefer.</li>
</ul>


### Question 1

In [97]:
normalized_ministerie = kvrdf.ministerie.value_counts().head(10) / sum(kvrdf.ministerie.value_counts())

print(normalized_ministerie)

 Justitie (JUS)                                                   0.118393
 Volksgezondheid, Welzijn en Sport (VWS)                          0.096730
 Buitenlandse Zaken (BUZA)                                        0.066056
 Verkeer en Waterstaat (VW)                                       0.052999
 Justitie                                                         0.049027
 Sociale Zaken en Werkgelegenheid (SZW)                           0.045276
 Onderwijs, Cultuur en Wetenschappen (OCW)                        0.043657
 Volkshuisvesting, Ruimtelijke Ordening en Milieubeheer (VROM)    0.036191
 Financiën (FIN)                                                  0.035308
 Volksgezondheid, Welzijn en Sport                                0.034977
dtype: float64


### Question 2